README.md


![](https://github.com/southpeak/iOS-tech-set/blob/master/images/Banner.png?raw=true)

**[English](https://github.com/southpeak/iOS-tech-set/blob/master/README_EN.md)**

## 初衷

16年某一天，在[__叶孤城](https://weibo.com/u/1438670852)的豪宅里，几个基友讨论了一些技术问题，颇有收获，回去整理了一下。于是就在微博上开了个话题`iOS知识小集`，来分享这些技术点。

`iOS知识小集`的初衷是希望用300字左右（外加代码和效果展示）来说明一个小知识点，这样读者可以在上下班路上，花个2分钟就能了解一个iOS开发的小知识。当然，短短几百字无法说清楚一个知识点，要详细的了解，还需要查阅更多的资料并且去实践一下。另外由于是零零散散地记录，所以成不了体系。

## 关于

最初是想每天发1~3条小`tip`出来，但发现整理一条`tip`没有我想象的快，另外自己放在iOS上的时间也越来越少，所以没能一直坚持，只是断断续续的在写一些，以致出现间断几个月的情况。到2017.10.28为止，自己大概发了180条出来。另外，一些朋友也会发一些小tip到这个话题里面。如果微博数据没有造假的话，这个话题已经有`1000w`的阅读量了。所以又萌生了把之前零零散散的小`tip`整理起来的念头，于是就在`Github`上开了这个`Repository`，方便大家阅读。整理的时间每周一次，把当周的知识小集归档进来。有些内容会过时，还请自行判断。

这个Repository里收录的除了我自己写的一些，也会有一些朋友发的小tip(主要是我转发到`iOS知识小集`话题中的)，这些我都会先征求对方的意见，并标明来源。若是作者有异义，可以随时联系我并删除，在此谢过。同时，有兴趣的朋友也可以一起来分享。

## 成员

一个人的力量毕竟有限，所以希望有更多的人来一起分享。感谢[Lefe_x](https://weibo.com/u/5953150140)童鞋，我俩一拍即合，想组个小团队，把这个分享做下去。如果你有想法、有责任心、有较强的自学能力和意愿，当然对我们小团队有兴趣的话，可以联系[南峰子_老驴](http://weibo.com/touristdiary) ，一起来尝试一下能把这个知识小集做成什么样子。

希望我们的队伍能慢慢壮大。

 <a href="https://weibo.com/touristdiary"><img style="border-radius: 30px" src="https://tva1.sinaimg.cn/crop.1.0.1366.1366.180/c5ff030ejw8f5bbc70i61j212011yq80.jpg" title="南峰子_老驴" width="60"/></a> | [南峰子_老驴](https://weibo.com/touristdiary) | <a href="https://weibo.com/u/5953150140"><img style="border-radius: 30px" src="https://tva4.sinaimg.cn/crop.8.0.1226.1226.180/006uSOiEjw8f9h4ihstq4j30yi0y2gnq.jpg" title="Lefe_x" width="60"/></a> | [Lefe_x](https://weibo.com/u/5953150140) 
------------- | ------------- | ------------- | -------------
<a href="https://weibo.com/VongLo"><img style="border-radius: 30px" src="https://tvax3.sinaimg.cn/crop.0.0.667.667.180/ba81ca29ly8fhu4meonedj20ij0ijgmh.jpg" title="Vong_HUST" width="60"/></a> | [Vong_HUST](https://weibo.com/VongLo) | <a href="https://weibo.com/soapyigu"><img style="border-radius: 30px" src="https://tva4.sinaimg.cn/crop.14.0.721.721.180/6cf34ee4jw8f8rdmtzzgmj20ku0k10t5.jpg" title="故胤道长" width="60"/></a> | [故胤道长](https://weibo.com/soapyigu)
<a href="https://weibo.com/517082456"><img style="border-radius: 30px" src="https://tva4.sinaimg.cn/crop.0.0.1242.1242.180/5fe18d75jw8evft9qcjh5j20yi0yigo5.jpg" title="高老师很忙" width="60"/></a> | [高老师很忙](https://weibo.com/517082456) | <a href="https://weibo.com/u/2293476232"><img style="border-radius: 30px" src="https://tvax1.sinaimg.cn/crop.6.0.737.737.180/88b3ab88ly8fnassmyvedj20ku0khgma.jpg" title="_拿破仑的_风火轮_" width="60"/></a> | [_拿破仑的_风火轮_](https://weibo.com/u/2293476232) 
<a href="http://weibo.com/halohily"><img style="border-radius: 30px" src="http://ww4.sinaimg.cn/mw690/d9ec7ffcjw8f8a753z961j20e80dp0t3.jpg" title="halohily" width="60"/></a> | [halohily](http://weibo.com/halohily) | <a href="https://weibo.com/kangzubin"><img style="border-radius: 30px" src="https://tva3.sinaimg.cn/crop.0.0.440.440.180/621b53aejw8ekybg28hxzj20c80c83z0.jpg" title="KANGZUBIN" width="60"/></a> | [KANGZUBIN](https://weibo.com/kangzubin) 
<a href="https://weibo.com/cimer"><img style="border-radius: 30px" src="https://tvax3.sinaimg.cn/crop.0.0.240.240.180/63fbed7aly8fgwp4qd9e4j206o06omx4.jpg" title="陈满iOS" width="60"/></a> | [陈满iOS](https://weibo.com/cimer)  | <a href="https://weibo.com/u/6603469503"><img style="border-radius: 30px" src="http://wx1.sinaimg.cn/mw690/007cTu4Dly8ftcswhzxpsj303o03oglg.jpg" title="这个汤圆没有馅" width="60"/></a> | [这个汤圆没有馅](https://weibo.com/u/6603469503) 
<a href="https://weibo.com/u/3951595216"><img style="border-radius: 30px" src="https://tvax3.sinaimg.cn/crop.0.0.750.750.180/eb888ed0ly8fzix3b96z7j20ku0kutad.jpg" title="NotFound--" width="60"/></a> | [NotFound--](https://weibo.com/u/3951595216)  

## 公众号与Medium

2018年1月份，在 [故胤道长](https://weibo.com/soapyigu) 的推动下，「 小集 」团队在 `Medium` 上开通了账号 [iOS Tech Set](https://medium.com/@iostechset)，并将小集的内容转换成英文，让小集走向了国际化。

2018年3月份，我们又开通了微信公众号`「 知识小集 」`，我们希望通过这个平台来分享更多的开发小知识。在这个平台上，我们会以文章的形式来呈现更多的内容。另外，我们会将每周微博 `#iOS知识小集#` 话题的内容整理成合集放到这个公众号，希望能带来更好的客户端阅读体验。欢迎大家关注我们的公众号。

<img src="https://raw.githubusercontent.com/iOS-Tips/iOS-tech-set/master/images/qrcode.jpg" title="知识小集" width="200"/>

## 最新内容
我们会定期整理微博上分享的内容，然后将最新的内容标题放在这里。更多内容可以查看每月的具体内容。

[2019 年 01 月](https://github.com/southpeak/iOS-tech-set/blob/master/2019/01.md)

* [1. Debug Memory Graph检查内存泄漏 【这个汤圆没有馅】](https://github.com/southpeak/iOS-tech-set/blob/master/2019/01.md#debug-memory-graph%E6%A3%80%E6%9F%A5%E5%86%85%E5%AD%98%E6%B3%84%E6%BC%8F)
* [2. 在 UILabel 中渲染 HTML 【南峰子】](https://github.com/southpeak/iOS-tech-set/blob/master/2019/01.md#%E5%9C%A8-uilabel-%E4%B8%AD%E6%B8%B2%E6%9F%93-html)
* [3. iOS App 异常捕获相互覆盖问题 【KANGZUBIN】](https://github.com/southpeak/iOS-tech-set/blob/master/2019/01.md#ios-app-%E5%BC%82%E5%B8%B8%E6%8D%95%E8%8E%B7%E7%9B%B8%E4%BA%92%E8%A6%86%E7%9B%96%E9%97%AE%E9%A2%98)
* [4. Aspects hook 类方法的正确姿势  【Vong】](https://github.com/southpeak/iOS-tech-set/blob/master/2019/01.md#aspects-hook-%E7%B1%BB%E6%96%B9%E6%B3%95%E7%9A%84%E6%AD%A3%E7%A1%AE%E5%A7%BF%E5%8A%BF-)
* [5. This block declaration is not a prototype 编译警告处理 【南峰子】](https://github.com/southpeak/iOS-tech-set/blob/master/2019/01.md#this-block-declaration-is-not-a-prototype-%E7%BC%96%E8%AF%91%E8%AD%A6%E5%91%8A%E5%A4%84%E7%90%86)
* [6. Framework 中混合编程时 umbrella header 设置注意事项 【南峰子】](https://github.com/southpeak/iOS-tech-set/blob/master/2019/01.md#framework-%E4%B8%AD%E6%B7%B7%E5%90%88%E7%BC%96%E7%A8%8B%E6%97%B6-umbrella-header-%E8%AE%BE%E7%BD%AE%E6%B3%A8%E6%84%8F%E4%BA%8B%E9%A1%B9)
* [7. 关于UIStackView的一个小知识点 【高老师很忙】](https://github.com/southpeak/iOS-tech-set/blob/master/2019/01.md#%E5%85%B3%E4%BA%8Euistackview%E7%9A%84%E4%B8%80%E4%B8%AA%E5%B0%8F%E7%9F%A5%E8%AF%86%E7%82%B9)
* [8. Xcode更新输入账号密码，账号却不是自己的 【Lefe】](https://github.com/southpeak/iOS-tech-set/blob/master/2019/01.md#xcode%E6%9B%B4%E6%96%B0%E8%BE%93%E5%85%A5%E8%B4%A6%E5%8F%B7%E5%AF%86%E7%A0%81%EF%BC%8C%E8%B4%A6%E5%8F%B7%E5%8D%B4%E4%B8%8D%E6%98%AF%E8%87%AA%E5%B7%B1%E7%9A%84)
* [9. Swift 中实现 synchronized 【南峰子】](https://github.com/southpeak/iOS-tech-set/blob/master/2019/01.md#swift-%E4%B8%AD%E5%AE%9E%E7%8E%B0-synchronized)
* [10. 点击cell不执行-[UITableView didSelectRowAtIndexPath:]方法的几种方式 【高老师很忙】](https://github.com/southpeak/iOS-tech-set/blob/master/2019/01.md#%E7%82%B9%E5%87%BBcell%E4%B8%8D%E6%89%A7%E8%A1%8C-%5Buitableview-didselectrowatindexpath%3A%5D%E6%96%B9%E6%B3%95%E7%9A%84%E5%87%A0%E7%A7%8D%E6%96%B9%E5%BC%8F)
* [11. Xcode 工程设置构建版本号自动递增 【KANGZUBIN】](https://github.com/southpeak/iOS-tech-set/blob/master/2019/01.md#xcode-%E5%B7%A5%E7%A8%8B%E8%AE%BE%E7%BD%AE%E6%9E%84%E5%BB%BA%E7%89%88%E6%9C%AC%E5%8F%B7%E8%87%AA%E5%8A%A8%E9%80%92%E5%A2%9E)
* [12. Command Not Found 【Lefe】](https://github.com/southpeak/iOS-tech-set/blob/master/2019/01.md#command-not-found)
* [13. NSScanner 过滤字符串 【这个汤圆没有馅】](https://github.com/southpeak/iOS-tech-set/blob/master/2019/01.md#nsscanner-%E8%BF%87%E6%BB%A4%E5%AD%97%E7%AC%A6%E4%B8%B2)
* [14. 针对 objc_exception_throw 的实用调试技巧 【Vong】](https://github.com/southpeak/iOS-tech-set/blob/master/2019/01.md#%E9%92%88%E5%AF%B9-objc_exception_throw-%E7%9A%84%E5%AE%9E%E7%94%A8%E8%B0%83%E8%AF%95%E6%8A%80%E5%B7%A7)
* [15. 聊聊 iPad 适配 【halohily】](https://github.com/southpeak/iOS-tech-set/blob/master/2019/01.md#%E8%81%8A%E8%81%8A-ipad-%E9%80%82%E9%85%8D)

[[2019]查看更多 [1] [2] [3] ... ➡️](https://github.com/southpeak/iOS-tech-set/blob/master/2019/目录.md)

[[2018]查看更多 [1] [2] [3] ... ➡️](https://github.com/southpeak/iOS-tech-set/blob/master/2018/目录.md)

[[2017]查看更多 [1] [2] [3] ... ➡️](https://github.com/southpeak/iOS-tech-set/blob/master/2017/目录.md)



news.md
## 44

• [Kuo: New ‘pro’ iPad models with mini-LED display coming Q4 2020 – Q1 2021, mini-LED MacBook after](https://9to5mac.com/2019/09/30/macbook-mini-led/)
• [Possible Nest Wifi router and beacon arrive at FCC with Thread support](https://9to5google.com/2019/09/30/google-nest-wifi-fcc/)
• [Google brings you on free VR Palace of Versailles tour](https://www.cnet.com/news/google-brings-you-on-free-vr-palace-of-versailles-tour/)
• [How to switch from Android to the iPhone](https://www.cnet.com/how-to/switching-from-android-to-the-iphone-11-heres-the-fastest-way/)
• [Working with Barcodes and QR Codes in Flutter](https://medium.com/flutter-community/working-with-barcodes-and-qr-codes-in-flutter-fbb3a2d4ede7)
• [Hooked on React's Array Destructuring Pattern](https://hackernoon.com/hooked-on-reacts-array-destructuring-pattern-ed32y3smn)
• [SwiftUI: Animating Timing Curves](https://www.objc.io/blog/2019/09/26/swiftui-animation-timing-curves/)
• [How to Implement Day/Night Mode in Your Android App](https://cobe.tech/blog/post/how-to-implement-day-night-mode-in-your-android-app)



script/main.py
# coding=UTF-8

import os
import tec_year_catagoy
import tec_teach_set
import tec_constant
import tec_readme

# 小集零时目录的标题
valid_file_names = set()


# 1. 检查小集的模版是否正确
def check_teachsets_valid(root_path):
    paths = os.listdir(root_path)

    is_all_valid = True

    for compent in paths:
        # 过滤掉隐藏文件
        if compent.startswith('.'):
            continue

        # 每个人的小集目录
        teach_set_path = os.path.join(root_path, compent)
        ext = os.path.splitext(teach_set_path)[1]
        if ext == '.md':
            valid_file_names.add(compent)
            # 检查小集内容是否合法
            if not tec_teach_set.is_valid_teachset_template(teach_set_path):
                if is_all_valid:
                    is_all_valid = False
                print(compent + ' : 的小集模版不正确')
        else:
            print('❌ : file must have extension')

    return is_all_valid


# 2.检查所有成员是否都提交了小集
def check_members_valid(root_path):
    subs = tec_constant.TEACHSET_MEMBERS() - valid_file_names
    if len(subs) > 0:
        for member_name in subs:
            print("检查到这些成员没有上传本周小集：" + member_name)
    return len(subs) > 0


# 查找小集目录下的文件
def parse_teach_set_file(path):

    if not check_teachsets_valid(path):
        return

    check_members_valid(path)

    # 3.所有小集检查成功，开始解析
    print('\n文件合法行检查成功，正在写入' + tec_constant.TEACHSET_LINE_CONSTANT() + '\n')
    for file_name in valid_file_names:
        aPath = os.path.join(path, file_name)
        print('开始处理 ' + file_name + tec_constant.TEACHSET_LINE_CONSTANT())
        tec_teach_set.write_teach_set_to_file(aPath)
        print('处理完成 ' + file_name + tec_constant.TEACHSET_LINE_CONSTANT())
        print('\n')

    # 保存小集年目录
    print('开始写入目录中 ' + tec_constant.TEACHSET_LINE_CONSTANT())
    tec_year_catagoy.save_catalog(valid_file_names)
    print('目录写入成功 ' + tec_constant.TEACHSET_LINE_CONSTANT())
    print('\n')

    # 更新 readme 文件
    print('开始写入 readme 文件中 ' + tec_constant.TEACHSET_LINE_CONSTANT())
    tec_readme.update_readme_file()
    print('readme 文件写入成功 ' + tec_constant.TEACHSET_LINE_CONSTANT())
    print('\n')

    print('本次小集已整理完成，提交前检查一遍')

# 脚本入口
if __name__ == '__main__':
    # 从小集的 temp 目录中读取文件
    parse_teach_set_file(tec_constant.TEACHSET_DESPATH_TEMP())

script/tec_readme.py
# coding=UTF-8

"""
【更新 readme 文件，readme 文件模版如下】

我们会定期整理微博上分享的内容，然后将最新的内容标题放在这里。更多内容可以查看每月的具体内容。
[2018 年 1 月](https://github.com/southpeak/iOS-tech-set/blob/master/2018/01.md)
* [动态加载Framework/Library【南峰子_老驴】](https://github.com/southpeak/iOS-tech-set/blob/master/2018/01.md)

[2018 年 2 月](https://github.com/southpeak/iOS-tech-set/blob/master/2018/02.md)
* [编译源文件的流程【南峰子_老驴】](https://github.com/southpeak/iOS-tech-set/blob/master/2018/02.md)

[[2017]查看更多 [1] [2] [3] ... ➡️](https://github.com/southpeak/iOS-tech-set/blob/master/2018/%E7%9B%AE%E5%BD%95.md)
[[2018]查看更多 [1] [2] [3] ... ➡️](https://github.com/southpeak/iOS-tech-set/blob/master/2018/%E7%9B%AE%E5%BD%95.md)

【模版说明】
1.第一行内容为：我们会定期整理微博上分享的内容，然后将最新的内容标题放在这里。更多内容可以查看每月的具体内容。
2.找出最新2个月的内容
3.最近两个月的内容可以跳转到对应的年目录
4.查看更多内容需要跳转到每年的具体年目录下

"""

import tec_constant
import tec_time
import tec_year_catagoy
import datetime
import os

def update_readme_file():

    # README 文件
    readme_file_path = tec_constant.TEACHSET_DESPATH() + '/' + tec_constant.TEACHSET_README_FILE_NAME()
    readme_file = open(readme_file_path, 'r')

    temp_file_path = tec_constant.TEACHSET_DESPATH() + '/' + 'tempreadme.md'
    temp_file = open(temp_file_path, 'w')
    locked = False
    have_update = False

    for index, line in enumerate(readme_file):
        if line.startswith('## 感谢'):
            locked = False

        if line.startswith('## 最新内容'):
            locked = True
        else:
            if not locked:
                temp_file.write(line)

        if locked and not have_update:
            have_update = True
            month = tec_time.current_month()
            year = tec_time.current_year()

            temp_file.write('## 最新内容')
            temp_file.write('\n')
            temp_file.write('我们会定期整理微博上分享的内容，然后将最新的内容标题放在这里。更多内容可以查看每月的具体内容。\n\n')

            catalogs = tec_year_catagoy.catalog_from_month(year, month)
            if len(catalogs) > 0:
                temp_file.write('[%s 年 %s 月](https://github.com/southpeak/iOS-tech-set/blob/master/%s/%s)\n' % (year, month, year, month + tec_constant.TEACHSET_FILE_EXTENSION()))

            for line in catalogs:
                temp_file.write(line + '\n')

            if len(catalogs) < 8:
                # 如果小于 10 条，取上个月的
                month_int = datetime.datetime.now().month
                year_int = datetime.datetime.now().year

                if month_int == 1:
                    month_str = '12'
                    year_str = str(year_int - 1)

                else:
                    year_str = str(year_int)
                    if month_int > 9:
                        month_str = str(month_int - 1)
                    else:
                        month_str = '0' + str(month_int - 1)

                temp_file.write('\n')
                temp_file.write('[%s 年 %s 月](https://github.com/southpeak/iOS-tech-set/blob/master/%s/%s)\n' % (
                year_str, month_str, year_str, month_str + tec_constant.TEACHSET_FILE_EXTENSION()))
                catalogs = tec_year_catagoy.catalog_from_month(year_str, month_str)
                for line in catalogs:
                    temp_file.write(line + '\n')

            # 查看更多
            current_year = datetime.datetime.now().year
            year_path = tec_constant.TEACHSET_DESPATH() + '/' + str(
            current_year) + '/' + tec_constant.TEACHSET_MONTH_FILE_NAME()
            temp_file.write('\n')
            while (os.path.exists(year_path)):
                temp_file.write(
                    '[[%s]查看更多 [1] [2] [3] ... ➡️](https://github.com/southpeak/iOS-tech-set/blob/master/%s/%s)\n\n' %(current_year, current_year, tec_constant.TEACHSET_MONTH_FILE_NAME()))
                current_year -= 1
                year_path = tec_constant.TEACHSET_DESPATH() + '/' + str(
                    current_year) + '/' + tec_constant.TEACHSET_MONTH_FILE_NAME()

    readme_file.close()
    temp_file.close()

    # 重命名，并删除旧的 readme 文件
    os.remove(readme_file_path)
    os.rename(temp_file_path, readme_file_path)

# 脚本入口
if __name__ == '__main__':
    # 从小集的 temp 目录中读取文件
    update_readme_file()

script/tec_teach_set.py
# coding=UTF-8


"""
【每条小集的模版】
iOS 关于音频播放调研
--------
**作者**: [Lefe_x](https://weibo.com/u/5953150140)

正文

【规则】
1. 第一行为标题
2. 第二行为： --------
3. 第三行为：**作者**: [微博名](微博链接)
4. 第四行为：空行
5. 第五行之后为正文
"""

import re
import os
import tec_time
import tec_constant

# 临时文件夹下小集标题的集合
def tempfile_teachset_titles_authors():
    root_path = tec_constant.TEACHSET_DESPATH_TEMP()
    paths = os.listdir(root_path)
    results = dict()
    # 区分作者
    index = 0
    for compent in paths:
        # 过滤掉隐藏文件
        if compent.startswith('.'):
            continue
        index += 1
        # 每个人的小集目录
        teach_set_path = os.path.join(root_path, compent)
        ext = os.path.splitext(teach_set_path)[1]
        if ext == '.md':
            teach_set_file = open(teach_set_path)
            lines = teach_set_file.readlines()
            if len(lines) > 0:
                # 作者 **作者**: [Lefe_x](https://weibo.com/u/5953150140)
                regx = '\[(.{0,})\]'
                matchs_author = re.findall(regx, lines[2])
                if matchs_author:
                    author = matchs_author[0]
                    results[author + '_' + str(index)] = lines[0].replace('\n','')

    return results


# 判断小集的内容模版是否正确
def is_valid_teachset_template(path):
    teach_set_file = open(path)
    lines = teach_set_file.readlines()
    if len(lines) < 3:
        teach_set_file.close()
        return False
    line1 = lines[1]
    line1 = line1.strip()
    regx = '^-{4,}'
    matchs_line1 = re.findall(regx, line1)

    author_line = lines[2]
    author_line = author_line.strip()
    regx = '^(\*\*作者\*\*)(:|：)'
    matchs_line2 = re.findall(regx, author_line)

    teach_set_file.close()

    return len(matchs_line1) > 0 and len(matchs_line2) > 0


# 每个文件中的标题集合
def all_teachset_titles(path):
    teach_set_file = open(path)
    lines = teach_set_file.readlines()
    titles = []

    if len(lines) == 0:
        teach_set_file.close()
        return titles

    for index, line in enumerate(lines):
        line = line.strip()
        regx = '^-{4,}'
        match_lines = re.findall(regx, line)
        if len(match_lines) > 0:
            if index != 0:
                title = lines[index - 1]
                titles.append(title)
    teach_set_file.close()
    return titles


# 把小集写入到当前月的内容中
def write_teach_set_to_file(path):
    year = tec_time.current_year()
    month = tec_time.current_month()
    year_file_path = tec_constant.TEACHSET_DESPATH() + '/' + year
    month_file_path = year_file_path + '/' + month + tec_constant.TEACHSET_FILE_EXTENSION()

    # 如果年文件夹不存在需要创建
    if not os.path.exists(year_file_path):
        os.mkdir(year_file_path)

    # 月文件，如果文件已经存在，文件需要写入空行
    if os.path.exists(month_file_path):
        # 检查是否已经写入过小集，需要用标题判断
        titles = all_teachset_titles(month_file_path)
        if len(titles) > 0:
            month_temp_file = open(path, 'r')
            lines = month_temp_file.readlines()
            if len(lines) > 0:
                f_line = lines[0]
                if f_line in titles:
                    print("该小集已经添加：" + f_line.strip('\n'))
                    month_temp_file.close()
                    return
            month_temp_file.close()

        month_file = open(month_file_path, 'a+')
        month_file.write('\n\n')

    else:
        # 月文件不存在需要，开头写入年月和空行
        month_file = open(month_file_path, 'a+')
        month_file.write('# ' + year + '.' + month)
        month_file.write('\n')

    # 遍历小集内容，写入小集中
    teactset_file = open(path)
    for index, line in enumerate(teactset_file):
        month_file.write(line)

    month_file.close()
    teactset_file.close()


# 脚本入口
if __name__ == '__main__':
    # 从小集的 temp 目录中读取文件
    path = '/Users/wangsuyan/Desktop/project/iOS-tech-set/TEMP/Lefe_x.md'
    titles =  all_teachset_titles(path)
    print('===============================')
    print(titles)

    month_file = open(path, 'r')
    llines = month_file.readlines()
    print('===============================')
    if len(llines):
        title = llines[0]
        if title in titles:
            print('Hello I am in')

script/tec_year_catagoy.py
# coding=UTF-8

"""
# 2017目录

## 01

* [1. Swift中lazy作惰性求值【Vong_HUST】](https://github.com/southpeak/iOS-tech-set/blob/master/2017/01.md#Swift中lazy作惰性求值)
* [2. Swift的stride操作【Vong_HUST】](https://github.com/southpeak/iOS-tech-set/blob/master/2017/01.md#Swift的stride操作)

"""

import os
import tec_time
import tec_constant
import tec_teach_set
import re
from urllib import quote

def github_url_encode(url):
    encode_title = url
    encode_title = encode_title.replace(' ', '-')
    encode_title = encode_title.replace('?', '')
    encode_title = encode_title.lower()
    encode_title = quote(encode_title)

    return encode_title

def catalog_from_month(year, month):
    results = []

    # 某年对应的目录文件
    year_catalog_file_path = tec_constant.TEACHSET_DESPATH() + '/' + year + '/' + tec_constant.TEACHSET_MONTH_FILE_NAME()
    if not os.path.exists(year_catalog_file_path):
        return results

    year_catalog_file = open(year_catalog_file_path, 'r')
    is_find = False
    for index, line in enumerate(year_catalog_file):
        # * [1. Swift中lazy作惰性求值](https://github.com/southpeak/iOS-tech-set/blob/master/2017/01.md#Swift中lazy作惰性求值)
        line = line.strip('\n')
        if is_find == False:
            month_regex = r'^(?:##)\s+([0-9]+)'
            month_match = re.findall(month_regex, line)
            if month_match:
                # 是月标题
                if month_match[0] == month:
                    is_find = True
                    continue

        if not is_find:
            continue

        # 下个月的标题，结束
        if line.startswith('##'):
            break

        results.append(line)

    year_catalog_file.close()
    return results

# 根据年目录查找标题
def current_month_catalog_titles():
    results = set()

    # 某年对应的目录文件
    year_catalog_file_path = tec_constant.TEACHSET_DESPATH() + '/' + tec_time.current_year() + '/' + tec_constant.TEACHSET_MONTH_FILE_NAME()
    if not os.path.exists(year_catalog_file_path):
        return results

    month = tec_time.current_month()
    year_catalog_file = open(year_catalog_file_path, 'r')
    is_find = False
    for index, line in enumerate(year_catalog_file):
        # * [1. Swift中lazy作惰性求值](https://github.com/southpeak/iOS-tech-set/blob/master/2017/01.md#Swift中lazy作惰性求值)
        line = line.strip()
        if is_find == False:
            month_regex = r'^(?:##)\s+([0-9]+)'
            month_match = re.findall(month_regex, line)
            if month_match:
                # 是月标题
                if month_match[0] == month:
                    is_find = True
                    continue

        if not is_find:
            continue

        # 下个月的标题，结束
        if line.startswith('##'):
            break

        regex = r'^(?:\*)\s*\[(.{0,})\]'
        title_match = re.findall(regex, line)
        for item in title_match:
            results.add(item)

    year_catalog_file.close()
    return results


def save_catalog(catalog):
    month = tec_time.current_month()
    year = tec_time.current_year()

    # 如果年文件夹不存在需要创建
    year_file_path = tec_constant.TEACHSET_DESPATH() + '/' + year
    if not os.path.exists(year_file_path):
        os.mkdir(year_file_path)

    # 某年对应的目录文件
    year_catalog_file_path = year_file_path + '/' + tec_constant.TEACHSET_MONTH_FILE_NAME()

    # 年目录文件
    if not os.path.exists(year_catalog_file_path):
        year_catalog_file = open(year_catalog_file_path, 'w')
        year_catalog_file.write('# ' + year + '目录')
        year_catalog_file.write('\n')
    else:
        year_catalog_file = open(year_catalog_file_path, 'a+')

    catalog_titles = current_month_catalog_titles()
    techset_dic = tec_teach_set.tempfile_teachset_titles_authors()

    if len(catalog_titles) == 0:
        year_catalog_file.write('\n## ' + month + '\n')

    index = len(catalog_titles)
    for (key, teachset_title) in techset_dic.items():
        author = key.split('_')[0]
        is_find = False
        # 防止重复插入
        for catalog in catalog_titles:
            if teachset_title in catalog:
                is_find = True
                break

        if is_find:
            print('该标题已经插入目录中，不需要插入：' + teachset_title)
            continue

        index += 1

        header = '* [%s. %s 【%s】]'%(str(index), teachset_title,author)
        url_source = 'https://github.com/southpeak/iOS-tech-set/blob/master/%s/%s#'%(year,month+tec_constant.TEACHSET_FILE_EXTENSION())
        encode_title = github_url_encode(teachset_title)

        catalog_one = '%s(%s%s)'%(header,url_source,encode_title)

        year_catalog_file.write('\n' + catalog_one)


# 根据年月录查找标题
def year_month_catalog_titles(year, month):
    results = []

    # 某年对应的目录文件
    year_catalog_file_path = tec_constant.TEACHSET_DESPATH() + '/' + year + '/' + tec_constant.TEACHSET_MONTH_FILE_NAME()
    if not os.path.exists(year_catalog_file_path):
        return results

    year_catalog_file = open(year_catalog_file_path, 'r')
    is_find = False
    for index, line in enumerate(year_catalog_file):
        print(line)
        # * [1. Swift中lazy作惰性求值](https://github.com/southpeak/iOS-tech-set/blob/master/2017/01.md#Swift中lazy作惰性求值)
        line = line.strip()
        if is_find == False:
            month_regex = r'^(?:##)\s+([0-9]+)'
            month_match = re.findall(month_regex, line)
            if month_match:
                # 是月标题
                if month_match[0] == month:
                    is_find = True
                    continue

        if not is_find:
            continue

        # 下个月的标题，结束
        if line.startswith('##'):
            break

        regex = r'^(?:\*)\s*\[(.{0,})\]'
        title_match = re.findall(regex, line)
        for item in title_match:
            results.append(item)

    year_catalog_file.close()
    return results

# 脚本入口
if __name__ == '__main__':
    template = '* [1. Lefe_x teach_test3 【Lefe_x】](https://github.com/southpeak/iOS-tech-set/blob/master/2018/03.md#Lefe_x teach_test3)'
    regex = r'^(?:\*)\s*\[(.{0,})\]'
    spaceMatch = re.findall(regex, template)
    for item in spaceMatch:
        print('space: ' + item)

    title = "## 12"
    title_regex = r'^(?:##)\s+([0-9]+)'
    titleMatch = re.findall(title_regex, title)
    for item in titleMatch:
        print(item)

    author = '**作者**: [Lefe_x](https://weibo.com/u/5953150140)'
    author_regx = '\[(.{0,})\]'
    authorMatch = re.findall(author_regx, author)
    for item in authorMatch:
        print(item)



    encode_title = 'iOS 你的APP中能藏的住秘密吗?'
    encode_result = 'ios-%E4%BD%A0%E7%9A%84app%E4%B8%AD%E8%83%BD%E8%97%8F%E7%9A%84%E4%BD%8F%E7%A7%98%E5%AF%86%E5%90%97'
                    #'ios-%E4%BD%A0%E7%9A%84app%E4%B8%AD%E8%83%BD%E8%97%8F%E7%9A%84%E4%BD%8F%E7%A7%98%E5%AF%86%E5%90%97'
                    #'ios-%E4%BD%A0%E7%9A%84app%E4%B8%AD%E8%83%BD%E8%97%8F%E7%9A%84%E4%BD%8F%E7%A7%98%E5%AF%86%E5%90%97%EF%BC%9F'
    encode_title = encode_title.replace(' ', '-')
    encode_title = encode_title.replace('?', '')
    encode_title = encode_title.lower()
    encode_title = quote(encode_title)
    if encode_result == encode_title:
        print('successful')
    else:
        print('falied')
    print('=======')
    print(encode_title)

    titles = year_month_catalog_titles('2018', '01')
    print(titles)